File: ABC211/D.js
'use strict';

function Main(input) {
	input = input.split('\n');
	const [N, M] = input[0].split(' ').map(Number);
	const mod = 10 ** 9 + 7;

	const g = [...Array(M + 1)].map(() => Array(0));

	for (let i = 0; i < M; i++) {
		let [a, b] = input[1 + i].split(' ').map(Number);
		a--, b--;
		g[a].push(b);
		g[b].push(a);
	}

	const dist = Array(N).fill(Infinity);
	const q = [];
	q.push(0);
	dist[0] = 0;
	const vs = [];
	while (q.length) {
		const v = q.pop();
		vs.push(v);
		for (let i = 0; i < g[v].length; i++) {
			if (dist[g[v][i]] !== Infinity) continue;
			dist[g[v][i]] = dist[v] + 1;
			q.push(g[v][i]);
		}
	}

	const dp = Array(N).fill(0);
	dp[0] = 1;
	console.log(dist);
	for (let i = 0; i < vs.length; i++) {
		const v = vs[i];
		for (let j = 0; j < g[v].length; j++) {
			const u = g[v][j];
			if (dist[u] === dist[v] + 1) {
				console.log(u, v);
				dp[u] += dp[v];
				dp[u] %= mod;
				console.log(dp);
			}
		}
	}

	console.log(dp);

	console.log(dp[N - 1]);
}

// Main(require('fs').readFileSync('/dev/stdin', 'utf8'));
Main(require('fs').readFileSync('./test.txt', 'utf8'));

File: ABC211/D.py
from collections import deque

N, M = map(int, input().split())
cities = [[] for _ in range(N)]

for _ in range(M):
    A, B = map(int, input().split())
    cities[A - 1].append(B - 1)
    cities[B - 1].append(A - 1)


que = [0]
dist = [None] * N
cnt = [0] * N
dist[0] = 0
cnt[0] = 1

count = 0

for v in que:
    for vv in cities[v]:
        if dist[vv] is None:
            dist[vv] = dist[v] + 1
            que.append(vv)
            cnt[vv] = cnt[v]
        elif dist[vv] == dist[v] + 1:
            cnt[vv] += cnt[v]

print(cnt[-1] % (10**9 + 7))
